### DISCUSSION TITLE
V2 Migration Doc

### DISCUSSION THREAD

**Discussion by vitalik** [pydantic/pydantic#5393] [Pydantic V2]
Hello,

Is there any v1>v2 migration documentation page available ? (if not I guess would be great to start one )

What should be used instead of the following v1 functions/classes:
 - pydantic.schema.model_schema
 - pydantic.utils.GetterDict

**Comment by samuelcolvin** [comment]
It's a work in progress, we have a first draft [here](https://docs.pydantic.dev/blog/pydantic-v2-alpha/#migration-guide), in addition, we're using [this project](https://github.com/orgs/pydantic/projects/2) to collect points to be added to the migration guide.

* `pydantic.schema.model_schema` is best replaced by just using `model.model_json_schema()`, for non model types, we have `AnalytsedType.json_schema()`
* `GetterDict` has been removed since extracting attributes from objects uses `pydantic-core`s `from_attributes` feature, it's probably best replaced by functional validators or the more powerful alias functionality, but depends what you were using it for?

**Reply by vitalik** [reply]
> * but depends what you were using it for?

In django-ninja GetterDict used for converting [some django-specific results to serilazible objects](https://github.com/vitalik/django-ninja/blob/bcd68fdf04575de78a0185eab31af5e59cda6c66/ninja/schema.py#L48-L86)

For example if some attribute is queryset - it's evaluated (queried from db) and a list returned, if's it's a file field - then string url of the file is returned 

Also [resolve_xxx methods](https://django-ninja.rest-framework.com/guides/response/#resolvers) that allowed users to create some "calculated" fields

**Reply by samuelcolvin** [reply]
make sense.

The fundamental divider is whether you know the field types when you build the core-schema - e.g. when you create the pydantic model.

If you know share of the queryset, you should be able to use aliases to take the URL from the file field, something like this

```py
class MyQuerysetModel(BaseModel):
    my_file_field: str = Field(alias=['my_file_field', 'url'])
```

This would fill `my_file_field` from `queryset.my_file_field.url`.

If that doesn't work, model validation accepts any `Mapping`, so you should be able to wrap your own implementation of `GetterDict` in something which passes `isinstance(x, Mapping)`.

But this will be slower and require more of your own code than just getting the pydantic model or equivalent to build the right pydantic-core schema to extract the right stuff from the queryset.

---

On "calculated" fields, we're still aiming to have computed fields ready for v2.

**Comment by samuelcolvin** [comment]
Worth adding `model_schema` has a 1-to-1 replacement:

https://github.com/pydantic/pydantic/blob/28e83965098c52536af2beb54dcd4da777a4fcd2/pydantic/json_schema.py#L1151

**Comment by vitalik** [comment]
Hi @samuelcolvin 

do I use alias corectly here to get url from an image ? 

```Python
class PictureModel(BaseModel):
    id: int
    title: str
    image: str = Field(alias=['image', 'url'])

    class Config:
        from_attributes = True


class Response(BaseModel):
    picture: PictureModel

    class Config:
        from_attributes = True


picture = Picture.objects.first() # query pic from django orm
obj = Response(picture=picture)
print(obj.model_dump())
```
gives

```
...
pydantic/_internal/_model_construction.py", line 189, in complete_model_class
    cls.__pydantic_validator__ = SchemaValidator(schema, core_config)
pydantic_core._pydantic_core.SchemaError: Invalid Schema:
model.schema.typed-dict.fields.image.serialization_alias
  Input should be a valid string [type=string_type, input_value=['image', 'url'], input_type=list]
```

> If that doesn't work, model validation accepts any Mapping, so you should be able to wrap your own implementation of GetterDict in something which passes isinstance(x, Mapping).

Do you have any example ?

**Reply by samuelcolvin** [reply]
You have it right, just that we have some bugs in how pydantic builds the pydantic-core schema for aliases, I'll get this fixed.

In the meantime, if you comment out [this line](https://github.com/pydantic/pydantic/blob/c206df4e49a13937621420232629c9ef35483b29/pydantic/_internal/_generate_schema.py#L536), and remove `field.alias` from [this line](https://github.com/pydantic/pydantic/blob/c206df4e49a13937621420232629c9ef35483b29/pydantic/_internal/_model_construction.py#L242), then the following works:

```py
import dataclasses

from pydantic import BaseModel, Field

class PictureModel(BaseModel):
    id: int
    title: str
    image: str = Field(alias=['image', 'url'])

    class Config:
        from_attributes = True


class Response(BaseModel):
    picture: PictureModel

    class Config:
        from_attributes = True


@dataclasses.dataclass
class Url:
    image: str
    url: str


@dataclasses.dataclass
class Picture:
    id: int
    title: str
    image: Url


picture = Picture(id=1, title='test', image=Url(image='image', url='https://example.com'))

obj = Response(picture=picture)
print(obj.model_dump())
#> {'picture': {'id': 1, 'title': 'test', 'image': 'https://example.com'}}
```

TL;DR - we need to fix this.

**Reply by samuelcolvin** [reply]
see #5426

**Comment by vitalik** [comment]
@samuelcolvin 

and a queryset part is also need some hook for conversion:

```Python
class Response(BaseModel):
    pictures: list[PictureModel] # LIST !

    class Config:
        from_attributes = True


qs = Picture.objects.all()
obj = Response(pictures=qs) # PASSING Queryset
print(obj.model_dump())

# gives:
# pictures
#  Input should be a valid list [type=list_type, input_value=<QuerySet [<Picture: id=1>]>, input_type=QuerySet]
```

ideally we should have some callback (like `GetterDict.get`, or like in json library `JsonEncoder.default`) where will be a way to convert some complex types to something dumpable with pydantic

**Reply by samuelcolvin** [reply]
for this you should use

https://github.com/pydantic/pydantic/blob/c206df4e49a13937621420232629c9ef35483b29/pydantic/decorators.py#L377

usage is basically the same as `validator` in pydantic v1, you'll find some examples in tests until we document it properly.

**Reply by vitalik** [reply]
I tried that - but both field_serializer and model_serializer are not called when I pass queryset:

```Python
class Response(BaseModel):
    pictures: list[PictureModel]

    class Config:
        from_attributes = True

    @field_serializer('pictures')
    def serialize_qs_field(v):
        print("calling serialize_qs_field", v, type(v))
        return list(v)

    @model_serializer
    def serialize_qs_model(self):
        print('calling serialize_qs_model', type(self.pictures))
        # self.pictures = list(self.pictures)
        return self

obj = Response(pictures=Picture.objects.all())

# pydantic/main.py", line 208, in __init__
#    __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
# pydantic_core._pydantic_core.ValidationError: 1 validation error for PictureModel
# pictures
#   Input should be a valid list [type=list_type, input_value=<QuerySet [<Picture: with image>]>, input_type=QuerySet]
```

**Reply by samuelcolvin** [reply]
Yes, as per the name their for serialising the model - e.g. `model_dump` and `model_dump_json`.

If you want to customise validation, please use `field_validator` instead.

**Reply by vitalik** [reply]
@samuelcolvin  well... @validator is kinda works, but it still does not cover like 66% of current django-ninja schemas capabilities

```Python
class Response(BaseModel):
    pictures: list[PictureModel]
    picture: PictureModel = None

    class Config:
        from_attributes = True

    @validator('*', pre=True)
    def do_validate(cls, v, values, **kw):
        print("calling do_validate")
        print('  v=', v)
        print('  values=', values)
        print('  kw=', kw)
        if isinstance(v, QuerySet):
            v = list(v)
        return v

# calling do_validate
#  v= <QuerySet [<Picture: id=1>]>
#  values= {}
#  kw= {}
``` 

the problem is that I do not have any context about the variable at this callback... maybe you can pass at least field name into kwargs ?

**Reply by samuelcolvin** [reply]
See do include `field_name`, it's on `info`, see [here](https://github.com/pydantic/pydantic-core/blob/cf543e2fa9640ffb1ad1b0ffbef02a9de73720b7/pydantic_core/core_schema.py#L107)

**Reply by vitalik** [reply]
sorry, do not understand last piece..

**Reply by dmontagu** [reply]
@vitalik I believe this is what Samuel is getting at:
```python
from pydantic import BaseModel, field_validator


class Response(BaseModel):
    x: int

    class Config:
        from_attributes = True

    @field_validator('x', mode='before')
    @classmethod
    def do_validate(cls, v, info):
        print("calling do_validate")
        print('  v=', v)
        print('  info=', info)
        print(repr(cls.model_fields[info.field_name]))
        return v

Response(x=1)

# calling do_validate
#   v= 1
#   info= ValidationInfo(config={'title': 'Response', 'from_attributes': True}, context=None, data={}, field_name='x')
# FieldInfo(annotation=int, required=True)
```
(I couldn't easily replicate with PictureModel/QuerySet but hopefully this is clear enough how it would work in principle.)

I'll note that we have removed support for `'*'` as the field name in `@field_validator` but could potentially re-add that if necessary.

**Reply by dmontagu** [reply]
> ideally we should have some callback (like GetterDict.get, or like in json library JsonEncoder.default) where will be a way to convert some complex types to something dumpable with pydantic

I agree that arbitrary iterables should work with list schemas (etc.), but looking at the pydantic core code it looks like this should in principle work now if you can call `iter(my_query_set)` and get an iterator. I certainly don't _promise_ that currently works, but I think we'd like to make it work. If you can help me set up an environment where you are getting this not-a-list error with django-ninja, I'm happy to help debug it.

**Reply by vitalik** [reply]
@dmontagu well I'm afraid that validator will not solve full django-ninja backward compatibility without nasty monkey patching - we really need some way of a proxy between ORM and pydantic model before pydantic will start accessing object attributes - see #5440

anyway - here you can find a test with queryset>list issue https://github.com/vitalik/pydantic-v2-djangoninja

**Comment by vitalik** [comment]
@dmontagu @samuelcolvin 

I think `@root_validator(pre=True)` might do most of the required tricks... 
As of now I'm stuck with `Model.__fields__` - where can I find it in v2 ?

**Reply by samuelcolvin** [reply]
Look at `Model.model_fields` - contains most of the same information.

**Reply by dmontagu** [reply]
@vitalik Just curious -- are you using the `@root_validator` to apply some sort of wrapper that will make it so that attribute accesses do the thing you want? (Essentially as a way to implement the custom hook you wanted to replace `GetterDict`?)

Just want to stay up to date on your progress, haven't had a chance to dive into the stuff you shared yesterday about GetterDict yet but it's not being ignored!

**Reply by vitalik** [reply]
@dmontagu yes, basically root_validator adds a wrap like GetterDict around orm object, and solves almost all features:

https://github.com/vitalik/pydantic-v2-djangoninja/blob/0144cb71bd05b1eb10598c0f6f9dc62e58ade596/testproj/schema.py#L135-L142

**Comment by vitalik** [comment]
@samuelcolvin 

> On "calculated" fields, we're still aiming to have computed fields ready for v2.

Do you already have idea how it will look like ? will it have access to `self` and `context` somehow ?

like

```Python
class Math(BaseModel):
    a: int
    b: int

    @decorator?
    def sum(self, context?) -> int:
        return self.a + self.b
```

**Comment by vitalik** [comment]
@dmontagu @samuelcolvin 

Few more questions

what's the alternative for these v1 imports?
`pydantic.typing import ForwardRef, evaluate_forwardref`


-----


definding custom datatype - this was working in v1:

```Python
from django.core.files.uploadedfile import UploadedFile as DjangoUploadedFile
from pydantic.fields import FieldInfo

__all__ = ["UploadedFile"]


class UploadedFile(DjangoUploadedFile):
    @classmethod
    def __get_validators__(cls: Type["UploadedFile"]) -> Iterable[Callable[..., Any]]:
        yield cls._validate

    @classmethod
    def _validate(cls: Type["UploadedFile"], v: Any) -> Any:
        if not isinstance(v, DjangoUploadedFile):
            raise ValueError(f"Expected UploadFile, received: {type(v)}")
        return v

    @classmethod
    def __modify_schema__(
        cls, field_schema: Dict[str, Any], field: Optional[FieldInfo]
    ) -> None:
        field_schema.update(type="string", format="binary")

```
giving:
```
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class 'ninja.files.UploadedFile'>. Setting `arbitrary_types_allowed=True` in the model_config may prevent this error.

For further information visit https://errors.pydantic.dev/2.0a2/u/schema-for-unknown-type
```
btw errors.pydantic.dev host seems not resolving

**Comment by Averylamp** [comment]
I was curious if there's a timeline on a broader v2 release.  I'm a new pydantic user and I've been trying to weigh stability of v2 vs v1 and extra support for dataclasses in v2.  
Ideally I could just start with v2 and assume that v2 will reach GA within the X couple (weeks?  months?).  I was curious if there's a place to go to understand the stability of the v2 project or if there's any opinions there?
Thanks a ton in advanced :).  I'm pretty excited by the project in general!

**Reply by samuelcolvin** [reply]
See #4887, I'll provide an update there now.

**Reply by Averylamp** [reply]
Thanks!!

**Comment by cmungall** [comment]
In addition to a migration doc it would be wonderful to have a doc on how to make your code work with both v1 and v2 (distinct from the v1 compatibility mode)

I'm currently struggling with a library with multiple independent transitive dependencies on pydantic. There are multiple communications happening with multiple developers of the various intermediate libraries and each has different views on migration timelines. Additionally the situation is confused by either unnecessary pinning, or not pinning where it is required. In my ecosystem the majority of uses of pydantic are quite plain and we can get away with package maintainers avoiding v2 features in the interim, and relaxing their pinning, which will buy time to migrate. However, a lot of developers are unclear what they should do.
